### ISSUE TITLE
False negatives with sourceType: module and ecmaVersion <= 5

### ISSUE THREAD

**Issue by timruffles** [eslint/eslint#9687]
**Tell us about your environment**

* **ESLint Version:**
v4.12.1
* **Node Version:**
8.4.0
* **npm Version:**
5.3.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
    "parserOptions": {
        "sourceType": "module",
        "ecmaVersion": 5,
    }
}
```

</details>

Running against this file, ecmaVersion is ignored (could also be '3'):
 
```js
// es6.js
const x = () => `hi`;
```

**What did you expect to happen?**

Either, I would see an error like the below, rejecting an unsupported config:

```bash
> eslint -c withoutModule.config.js es6.js
Incompatible config: sourceType: module cannot be used with ecmaVersion <= 5
```

or the (admittedly niche*) use-case of ES6 modules with ES5 syntax would be supported, and I'd see errors:

```bash
> eslint -c withoutModule.config.js es6.js
es6.js
  1:1  error  Parsing error: The keyword 'const' is reserved

✖ 1 problem (1 error, 0 warnings)
```

which I get correctly with this config:

```js
// withoutModule.config.js
module.exports = {
    "parserOptions": {
        "ecmaVersion": 5,
    }
}
```

*niche: I was using rollup with some old ES5 code, and only wanted to modify the code to add imports. I tried the above, and thought "cool, it's working", but actually it was silently ignoring the ecmaVersion.

**What actually happened? Please include the actual, raw output from ESLint.**

No errors are reported:

```bash
> eslint es6.js
> echo $?
0
```

**Comment by not-an-aardvark** [comment]
Thanks for the report. I agree with you that the current interface is a bit confusing. We don't currently support parsing imports without parsing the rest of ES6, and as you've noticed, `sourceType: module` implicitly enables ES6 parsing.

Having said that, I'm not sure it would be worthwhile to introduce a error for this case, because there are probably a significant number of users relying on this behavior. We try to avoid breaking peoples' builds unnecessarily, even in major releases. (It's probably uncommon for users to do something like `{ ecmaVersion: 5, sourceType: module }` explicitly, but `ecmaVersion: 5` is the default, so anyone currently using `{ sourceType: module }` and expecting modules to be parsed is relying on this behavior.)

**Comment by timruffles** [comment]
@not-an-aardvark I would say breaking builds here is desirable. If people have this configuration, they're unambiguously asking explicitly for ES5 syntax + imports. What they're getting is ES6, with no warning that they're using ES6 syntax. That's dangerous.

**Comment by not-an-aardvark** [comment]
To clarify: I would agree that it would be good to report an error if the user puts `{ ecmaVersion: 5, sourceType: module }` in their configuration. However, I don't think it would be a good idea to report an error if the user only puts `{ sourceType: module }` in their configuration, because the user hasn't explicitly specified `ecmaVersion: 5`, and in most cases they probably want ES6 if they're using modules.

**Comment by lapo-luchini** [comment]
Using WebPack it's possible to use ES6 modules without "really" using them in the output, so checking againt ES5 (or even ES3) would make sense.

**Comment by Avaq** [comment]
I would really like to see support for es5 syntax with es6 modules. For me as a library author, this is a very common use-case, and I expect it to remain so until nobody cares about es5 support anymore.

- https://github.com/fluture-js/Fluture/blob/c4d22b0546cd97de6ac62b1daa5d2b9951a997c3/.eslintrc.json#L6-L7
- https://github.com/wearereasonablepeople/react-collect/blob/b0a9f8f156c2a0fbca40c793e117cabc661a8ca1/.eslintrc.json#L5-L6
- https://github.com/wearereasonablepeople/warped-components/blob/32149d92d82c454c84306aa11e2eeb6310aa3087/.eslintrc.json#L5-L6

I've been setting things up like this, hoping that support would land "any time now" for almost a year.

My reason for settings things up this way is because it gives me the best of all worlds:

1. My source is modular and organised.
1. My source is executable without a compilation step by `<script type="module">` and `node --experimental-modules`.
1. My source is "tree-shakable" and easily concatenated into a single file by tools such as Rollup
1. When concatenated, no further compilation is needed to target older environments. This in particular is the accepted standard among front-end developers. Nobody likes to consume libraries that have to be compiled.

**Comment by nzakas** [comment]
I think having an error for this use case makes sense, as it's confusing.

**TSC Summary:** Right now if you use `ecmaVersion: 5` with `sourceType: module`, we silently ignore the `ecmaVersion` setting and upgrade to `ecmaVersion: 6`. The end user can't tell that has happened, which may lead to confusion.

**TSC Question:** Shall we throw an error when `ecmaVersion` is less than 6 and `sourceType: module` is used? This would be a breaking change.

**Comment by kaicataldo** [comment]
**TSC Resolution:** This is accepted as a breaking change. Implementation details can be discussed on this issue.

**Comment by platinumazure** [comment]
@eslint/eslint-tsc Where do we feel we are at, in terms of the complexity of open questions on this issue? Should this be discussed in an RFC, or do we think we have a generally good idea of the direction here?

**Comment by nzakas** [comment]
I think the agreement was to throw an error when `sourceType: module` and `ecmaVersion` is less than 6 rather than silently changing it. I'm not sure we need much else to move forward.

**Comment by platinumazure** [comment]
@not-an-aardvark @aladdin-add Is this resolved by the espree changes?

**Comment by not-an-aardvark** [comment]
It might be a good idea to do a major (pre)release of espree and update to that version in ESLint before we close this.

**Comment by jefferyto** [comment]
I know I'm extremely late to this conversation, but I felt I should mention a use case for es5 syntax checking with es6 modules (in addition to the ones already mentioned):

[Rollup](https://rollupjs.org/) uses es6 modules exclusively but doesn't require any other es6 features. If I want to use eslint with Rollup, I am now essentially forced into using es6 validation.

For most forward-looking projects, this perhaps isn't a problem. But if I am writing a library for current/legacy browsers and want to author es5 only, then I only have two options:

* Just remember to not use es6 features
* Add babel to transpile any accidental es6 code

Neither of these options appeal to me because I want to author es5 code and want eslint to throw errors when that isn't the case.

**Comment by mysticatea** [comment]
@jefferyto Previously, ESLint had set `ecmaVersion: 6` implicitly automatically if `sourceType: "module"` was given but `ecmaVersion` was smaller than 6. This PR just removed that implicit behavior. You have recognized that it had not worked as your expectation now :)

You can use [eslint-plugin-es](https://mysticatea.github.io/eslint-plugin-es/) to warn arbitrary syntaxes individually.

**Comment by jefferyto** [comment]
@mysticatea Thanks for the pointer to eslint-plugin-es - so now I set `ecmaVersion: 6`, then use the `plugin:es/no-2015` preset, then turn off `es/no-modules`. This is... madness :joy:
